T112038261 reduce test runtime

coleholcomb opened the pull request:
## What does this PR do?
Decreased CPU test suite runtime in Meta internal testing utility. This was requested via a Meta task filed by @dianaml0.

The changes include:

- decreasing the number of parameterized test cases in some of the highest cardinality test cases
- reducing tensor sizes in some of the slowest running tests (several minute runtime in Sandcastle -> few seconds)
- small number of changes to reduce inefficiencies in the code involving itertools.product()

## Before submitting

- [🙃 ] Did you have fun?
- [X] Did you read the [contributor guideline](https://github.com/facebookresearch/xformers/blob/master/CONTRIBUTING.md)?
- [-] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  This was requested via a Meta task.
- [-] Did you make sure to update the docs?
  - [X] N/A
- [-] Did you write any new necessary tests?
  - [X] N/A
- [-] Did you update the [changelog](https://github.com/facebookresearch/xformers/blob/master/CHANGELOG.md)? (if needed)
  - [X] N/A


## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

dianaml0 commented on tests/test_attentions.py:
```diff
@@ -81,7 +81,7 @@ def noop(x):
     return multi_head
 
 
-@pytest.mark.parametrize("attn_dropout", [0.0, 0.3])
```
I think it would still be valuable here to test the case of no dropout

dianaml0 commented on tests/test_nystrom_attention.py:
```diff
@@ -13,14 +13,14 @@
 
 @pytest.mark.parametrize("pinverse_original_init", [True, False])
 @pytest.mark.parametrize("use_razavi_pinverse", [True, False])
-@pytest.mark.parametrize("num_landmarks", [30, 33, 905])
```
I think it's still valuable to test this because we test different code paths, one where the sequence length is divisible by the number of landmarks and one where it isn't

coleholcomb replied:
Currently I have [33, 905], would [30, 905] be preferable? This is a somewhat slow test, so I'm trying to reduce the test cases.

coleholcomb replied:
Based on this feedback, I refactored this test into two separate tests. The "close_to_sdp" test includes are three parameters and their original cases, the other test has fixed parameters.

coleholcomb replied:
By the way, is it intentional that the logic is repeated within the test_close_to_sdp() function?

blefaudeux commented on tests/test_attentions.py:
```diff
@@ -81,7 +81,7 @@ def noop(x):
 
 
-@pytest.mark.parametrize("attn_dropout", [0.0, 0.3])
+@pytest.mark.parametrize("attn_dropout", [0.3])
```
0 and 0.3 are different codepaths (0 will skip dropout completely for most implementations), I think we should keep both for safety reasons

blefaudeux commented on tests/test_attentions.py:
```diff
@@ -103,7 +103,7 @@ def test_order_invariance(
 
     # Check that a shuffled input produces the same results
-    seqs = [SEQ, SEQ - 16] if (attention_name != "blocksparse") else [SEQ]
+    seqs = [SEQ, SEQ // 2] if (attention_name != "blocksparse") else [SEQ]
```
yop good point, that works !

blefaudeux commented on tests/test_compositional_attention.py:
```diff
@@ -29,10 +29,10 @@
 
 
-@pytest.mark.parametrize("attn_dropout", [0.0, 0.3])
+@pytest.mark.parametrize("attn_dropout", [0.3])
```
same as before, we should keep 0 because different codepath down the line ([see](https://github.com/facebookresearch/xformers/blob/main/xformers/triton/dropout.py#L171) for instance)

blefaudeux commented on tests/test_global_attention.py:
```diff
@@ -31,7 +31,7 @@ def test_ratio(global_attention_ratio: float):
         # results as the monolithic dense scaled_dot_product
         mask = config["attention_query_mask"][:, 0]
-        torch.allclose(r_global[:, mask, :], r_dense[:, mask, :])
+        assert torch.allclose(r_global[:, mask, :], r_dense[:, mask, :])
```
wops, thanks for the catch here

blefaudeux reviewed (commented):
LGTM except for the dropout, skipping the dropout call when the prob is 0 is an easy optimization that a lot of implementation use (including our own triton layers), I would put this back. Other than that this is great for me, thanks a bunch @coleholcomb !

coleholcomb commented:
I refactored the nystrom attention test into two separate tests. The "close_to_sdp" test includes are three parameters and their original cases, the other test has fixed parameters.

I also returned the dropout = 0.0 cases, which has slightly increased the test suite runtime by about a minute in the buck test. Since some of these tests have many parameters (originally 2^9 test cases), it may be worthwhile to test the dropout code paths independently with new tests in the future.

Also want to note that there is another torch.allclose() with no assert statement on line 105 of test_compositional_attention.py. However, when I added the assert, the test failed even before I had changed the test parameters. May be worth examining.

blefaudeux commented:
> I refactored the nystrom attention test into two separate tests. The "close_to_sdp" test includes are three parameters and their original cases, the other test has fixed parameters.
> 
> I also returned the dropout = 0.0 cases, which has slightly increased the test suite runtime by about a minute in the buck test. Since some of these tests have many parameters (originally 2^9 test cases), it may be worthwhile to test the dropout code paths independently with new tests in the future.
> 
> Also want to note that there is another torch.allclose() with no assert statement on line 105 of test_compositional_attention.py. However, when I added the assert, the test failed even before I had changed the test parameters. May be worth examining.

thanks, I agree on the dropout case, maybe that it could be out of the combinatorics but tested on the side, that's also fine by me ! You can comment out the compositional_test then, it's on me, I'll file an issue right now

blefaudeux reviewed (approved):
LGTM, thanks again @coleholcomb !
